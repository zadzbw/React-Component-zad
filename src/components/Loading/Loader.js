/**
 * Created by zad on 17/1/13.
 */
import React from 'react';
import classNames from 'classnames';
import isOneOf from '../../utils/isOneOf';

const loaderPrefix = 'zad-loader';
const _type = {
  size: ['small', 'large']
};

export default function ({
  size = undefined,
  tip = '',
  loading,
}) {
  const sizeSuffix = ({
    large: 'lg',
    small: 'sm',
  })[size];
  const loadingClass = classNames(loaderPrefix, {
    [`${loaderPrefix}-${sizeSuffix}`]: size && isOneOf(_type.size, size),
    [`${loaderPrefix}-show-tip`]: !!tip,
    [`${loaderPrefix}-is-loading`]: loading,
  });

  return (
    <div className={loadingClass}>
      <div className={`${loaderPrefix}-entity`}>
        <svg className={`${loaderPrefix}-svg`} viewBox="25 25 50 50">
          <circle className={`${loaderPrefix}-circle`} cx="50" cy="50" r="20" fill="none"/>
        </svg>
      </div>
      {!!tip && <div className={`${loaderPrefix}-tip`}>{tip}</div>}
    </div>
  );
}
