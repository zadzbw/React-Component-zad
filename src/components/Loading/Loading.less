@import "../../style/variables";

.zad-loader {
  color: @text-color-secondary;
  text-align: center;
  vertical-align: middle;
  display: none;
  opacity: 0;

  &-is-loading {
    opacity: 1;
    position: static;
    display: inline-block;
  }

  &-entity {
    display: inline-block;
    width: @loader-base-size;
    height: @loader-base-size;
  }

  &-sm .zad-loader-entity {
    width: @loader-sm-size;
    height: @loader-sm-size;
  }

  &-lg .zad-loader-entity {
    width: @loader-lg-size;
    height: @loader-lg-size;
  }

  &-svg {
    animation: rotate 2.5s linear infinite;
    height: 100%;
    transform-origin: center center;
    width: 100%;
  }

  &-circle {
    stroke-width: 5;
    stroke-miterlimit: 10;
    stroke-dasharray: 1, 151;
    stroke-dashoffset: 0;
    stroke-linecap: round;
    animation: loader-dash 2s @ease-in-out infinite, loader-color 7s @ease-in-out infinite;
  }
}

.zad-loading {
  &-has-children {
    position: relative;

    .zad-loader {
      position: absolute;
      height: 100%;
      width: 100%;
      z-index: 5;

      &-entity {
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
      }

      &-show-tip .zad-loader-entity {
        transform: translate(-50%, -50%) translateY(-10px);
      }

      &-tip {
        position: absolute;
        top: 50%;
        width: 100%;
        user-select: none;
        animation: tip-color 7s @ease-in-out infinite;
        transform: translateY(@loader-base-size / 2 - 6);
      }

      &-sm .zad-loader-tip {
        transform: translateY(@loader-sm-size / 2 - 6);
      }

      &-lg .zad-loader-tip {
        transform: translateY(@loader-lg-size / 2 - 6);
      }
    }
  }

  &-container {
    position: relative;
    transition: all 0.3s @ease-in-out;
  }

  &-blur {
    opacity: 0.7;
    filter: blur(1px);
    &:after {
      content: '';
      position: absolute;
      left: 0;
      right: 0;
      top: 0;
      bottom: 0;
      background: transparent;
    }
  }
}

@keyframes rotate {
  100% {
    transform: rotate(360deg);
  }
}

@keyframes loader-dash {
  0% {
    stroke-dasharray: 1, 151;
    stroke-dashoffset: 0;
  }
  50% {
    stroke-dasharray: 89, 151;
    stroke-dashoffset: -35px;
  }
  100% {
    stroke-dasharray: 89, 151;
    stroke-dashoffset: -124px;
  }
}

@keyframes loader-color {
  0%, 100% {
    stroke: @highlight-color;
  }
  25% {
    stroke: @primary-color;
  }
  50% {
    stroke: @success-color;
  }
  75% {
    stroke: @warning-color;
  }
}

@keyframes tip-color {
  0%, 100% {
    color: @highlight-color;
  }
  25% {
    color: @primary-color;
  }
  50% {
    color: @success-color;
  }
  75% {
    color: @warning-color;
  }
}
