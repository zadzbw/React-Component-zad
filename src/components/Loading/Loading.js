/**
 * Created by zad on 17/1/12.
 */
import React, {PropTypes, Component} from 'react';
import classNames from 'classnames';
import Animate from 'rc-animate';
import Loader from './Loader';

import './Loading.less';

const loadingPrefix = 'zad-loading';

export default class Loading extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: this.props.loading
    };
    this.debounceLoading = this.debounceLoading.bind(this);
  }

  static propTypes = {
    size: PropTypes.oneOf(['small', 'default', 'large']),
    loading: PropTypes.bool,
    tip: PropTypes.string,
    delay: PropTypes.number,
  };

  static defaultProps = {
    loading: true,
    delay: undefined,
  };

  // 对改变loading状态进行去抖
  debounceLoading(nextProps) {
    const prevLoading = this.props.loading;
    const nextLoading = nextProps.loading;
    const {delay} = this.props;

    if (this.debounceTimer) {
      clearTimeout(this.debounceTimer);
    }
    // 如果设定了delay且前后状态不一致
    if (delay && !isNaN(+delay) && prevLoading !== nextLoading) {
      this.debounceTimer = setTimeout(() => this.setState({loading: nextLoading}), delay);
    } else {
      this.setState({loading: nextLoading});
    }
  }

  componentWillReceiveProps(nextProps) {
    this.debounceLoading(nextProps);
  }

  componentWillUnmount() {
    if (this.debounceTimer) {
      clearTimeout(this.debounceTimer);
      this.debounceTimer = null;
    }
  }

  render() {
    const {size, tip, children} = this.props;
    const {loading} = this.state;

    if (children) {
      const containerClass = classNames({
        [`${loadingPrefix}-container`]: true,
        [`${loadingPrefix}-blur`]: loading,
      });

      const loader = loading ? (
        <div key={'loader'}>
          <Loader size={size} tip={tip} loading={loading}/>
        </div>
      ) : null;

      return (
        <Animate className={`${loadingPrefix}-has-children`} transitionName="fade" component="div">
          {loader}
          <div key="container" className={containerClass}>
            {children}
          </div>
        </Animate>
      );
    }

    return (
      <Loader size={size} tip={tip} loading={loading}/>
    );
  }
}
