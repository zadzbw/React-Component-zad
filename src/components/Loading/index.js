/**
 * Created by zad on 17/1/12.
 */
import Loading from './Loading';

export default Loading;
