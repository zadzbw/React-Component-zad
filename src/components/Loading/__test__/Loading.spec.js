/**
 * Created by zad on 17/1/16.
 */
import React from 'react';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import Loading from '../../Loading';
import Loader from '../../Loading/Loader';

describe('Loading Test', () => {
  it('normal', () => {
    const wrapper = shallow(
      <Loading size={'other'}/>
    );

    expect(wrapper.shallow().hasClass('zad-loader-is-loading')).toBeTruthy();
    expect(wrapper.shallow().hasClass('zad-loader-show-tip')).toBeFalsy();
    expect(wrapper.shallow().hasClass('zad-loader-sm')).toBeFalsy();
    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });

  it('size', () => {
    const wrapper = shallow(
      <Loading size={'small'}/>
    );

    expect(wrapper.shallow().hasClass('zad-loader-is-loading')).toBeTruthy();
    expect(wrapper.shallow().hasClass('zad-loader-show-tip')).toBeFalsy();
    expect(wrapper.shallow().hasClass('zad-loader-sm')).toBeTruthy();
    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });

  it('has tip', () => {
    const wrapper = shallow(
      <Loading tip={'hello'}/>
    );

    expect(wrapper.shallow().hasClass('zad-loader-is-loading')).toBeTruthy();
    expect(wrapper.shallow().hasClass('zad-loader-show-tip')).toBeTruthy();
    expect(wrapper.shallow().hasClass('zad-loader-sm')).toBeFalsy();
    expect(wrapper.shallow().find('.zad-loader-tip').text()).toEqual('hello');
    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });

  it('not loading', () => {
    const wrapper = shallow(
      <Loading loading={false}/>
    );

    expect(wrapper.shallow().hasClass('zad-loader-is-loading')).toBeFalsy();
    expect(wrapper.shallow().hasClass('zad-loader-show-tip')).toBeFalsy();
    expect(wrapper.shallow().hasClass('zad-loader-sm')).toBeFalsy();
    expect(wrapper.shallow().find('.zad-loader-tip').isEmpty()).toBeTruthy();
    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });


  it('has children', () => {
    const wrapper = shallow(
      <Loading>
        <div className="test">
          test
        </div>
      </Loading>
    );

    expect(wrapper.find(Loader).props().loading).toEqual(true);
    expect(wrapper.find('.zad-loading-container').hasClass('zad-loading-blur')).toBeTruthy();
    expect(wrapper.find('.zad-loading-container').text()).toEqual('test');
    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });
});
