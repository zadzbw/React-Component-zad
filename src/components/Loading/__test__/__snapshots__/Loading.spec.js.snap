exports[`Loading Test has children 1`] = `
<div
  className="zad-loading-has-children">
  <AnimateChild
    animation={Object {}}
    transitionAppear={false}
    transitionEnter={true}
    transitionLeave={true}
    transitionName="fade">
    <div>
      <Loader
        loading={true} />
    </div>
  </AnimateChild>
  <AnimateChild
    animation={Object {}}
    transitionAppear={false}
    transitionEnter={true}
    transitionLeave={true}
    transitionName="fade">
    <div
      className="zad-loading-container zad-loading-blur">
      <div
        className="test">
        test
      </div>
    </div>
  </AnimateChild>
</div>
`;

exports[`Loading Test has tip 1`] = `
<div
  className="zad-loader zad-loader-show-tip zad-loader-is-loading">
  <div
    className="zad-loader-entity">
    <svg
      className="zad-loader-svg"
      viewBox="25 25 50 50">
      <circle
        className="zad-loader-circle"
        cx="50"
        cy="50"
        fill="none"
        r="20" />
    </svg>
  </div>
  <div
    className="zad-loader-tip">
    hello
  </div>
</div>
`;

exports[`Loading Test normal 1`] = `
<div
  className="zad-loader zad-loader-is-loading">
  <div
    className="zad-loader-entity">
    <svg
      className="zad-loader-svg"
      viewBox="25 25 50 50">
      <circle
        className="zad-loader-circle"
        cx="50"
        cy="50"
        fill="none"
        r="20" />
    </svg>
  </div>
</div>
`;

exports[`Loading Test not loading 1`] = `
<div
  className="zad-loader">
  <div
    className="zad-loader-entity">
    <svg
      className="zad-loader-svg"
      viewBox="25 25 50 50">
      <circle
        className="zad-loader-circle"
        cx="50"
        cy="50"
        fill="none"
        r="20" />
    </svg>
  </div>
</div>
`;

exports[`Loading Test size 1`] = `
<div
  className="zad-loader zad-loader-sm zad-loader-is-loading">
  <div
    className="zad-loader-entity">
    <svg
      className="zad-loader-svg"
      viewBox="25 25 50 50">
      <circle
        className="zad-loader-circle"
        cx="50"
        cy="50"
        fill="none"
        r="20" />
    </svg>
  </div>
</div>
`;
